nodejs/aws-infra/experimental/image.ts

```diff
@@ -0,0 +1,206 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as aws from "@pulumi/aws";
+import * as docker from "@pulumi/docker";
+import * as pulumi from "@pulumi/pulumi";
+
+import * as utils from "../utils";
+
+import * as mod from ".";
+
+export type ImageEnvironment = Record<string, pulumi.Input<string>>;
+
+export interface IImageProvider {
+    image(name: string, parent: pulumi.Resource): pulumi.Input<string>;
+    environment(name: string, parent: pulumi.Resource): pulumi.Input<aws.ecs.KeyValuePair[]>;
+}
+
+export abstract class ImageProvider implements IImageProvider {
```

Comment:
providers are paired with a concrete factory class that can be used to easily create instances for patterns we want to support.  but users can always supply just an impl of hte interface.  we never depend on downcasting to the concrete types.